samcli/lib/utils/resource_trigger.py

```diff
@@ -103,22 +103,28 @@ def get_dir_path_handler(dir_path: Path, ignore_regexes: Optional[List[str]] = N
 
 class TemplateTrigger(ResourceTrigger):
     _template_file: str
+    _stack_name: str
     _on_template_change: OnChangeCallback
     _validator: DefinitionValidator
 
-    def __init__(self, template_file: str, on_template_change: OnChangeCallback) -> None:
+    def __init__(self, template_file: str, stack_name: str, on_template_change: OnChangeCallback) -> None:
         """
         Parameters
         ----------
         template_file : str
             Template file to be watched
+        stack_name: str
+            Stack name of the template
         on_template_change : OnChangeCallback
             Callback when template changes
         """
         super().__init__()
         self._template_file = template_file
+        self._stack_name = stack_name
         self._on_template_change = on_template_change
         self._validator = DefinitionValidator(Path(self._template_file))
+        if not self._validator.validate():
```

Comment:
Why we need to call `validate` here since we didn't need it earlier?

Reply:
Because I don't want us to create a code trigger when the template is not valid. Since for nested stacks we have multiple template triggers it is easier to track it this way by providing information on stack + template name before other errors occur. Since watch is still in beta I wouldn't consider this as a breaking change either.

Reply:
This will break a use case where user will have an invalid template and start `sam sync --watch` and that template file won't be watched. If user fixes their template and expects sam to sync it, it won't happen. So I think we shouldn't be failing for invalid templates here.

Reply:
@mndeveci Yes I already added warning message on any template validation failure. The executions happen in an order from queuing the infra sync to creating the trigger. I do agree that catching the failure here and not creating the trigger is too late since the failed deployment should have already happened. I will take this raw validation part out of the initialization and print a warning message when a template is invalid